File: implementations/vb/Loom.Engine/Context/RagManager.vb
Imports Loom.Core.Enums
Imports Loom.Core.Models

Namespace Context
    ''' <summary>
    ''' Ingestion, sort chunks by score and ensures the MaxTokens limit is respected
    ''' </summary>
    Public Class RagManager

        Private ReadOnly _invocation As LlmInvocation

        Public Sub New(invocation As LlmInvocation)
            _invocation = invocation
        End Sub

        ''' <summary>
        ''' Adds a set of retrieved results to the RAG context
        ''' </summary>
        ''' <param name = "chunks" > The retrieved text chunks</param>
        ''' <param name = "clearExisting" > If True, clears the previous context</param>
        Public Sub AddResults(chunks As IEnumerable(Of RagChunk), Optional clearExisting As Boolean = False)

            If clearExisting Then _invocation.Rag.RetrievedChunks.Clear()

            Dim filtered As New List(Of RagChunk)

            For Each c In chunks
                If Not String.IsNullOrWhiteSpace(c.Text) Then
                    filtered.Add(c)
                End If
            Next

            Dim validChunks = filtered.OrderByDescending(Function(c) c.Score)

            _invocation.Rag.RetrievedChunks.AddRange(validChunks)

        End Sub

        ''' <summary>
        ''' Set strategy injection at runtime
        ''' </summary>
        ''' <param name="strategy"></param>
        Public Sub SetStrategy(strategy As InjectionStrategy)
            _invocation.Rag.Strategy = strategy
        End Sub

        ''' <summary>
        ''' Check if RAG context has sufficient data
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property HasContext As Boolean
            Get
                Return _invocation.Rag.RetrievedChunks.Any()
            End Get
        End Property

    End Class
End Namespace

File: implementations/csharp/src/Loom.Engine/Context/RagManager.cs
using Loom.Core.Enums;
using Loom.Core.Models;
using System.Collections.Generic;
using System.Linq;

namespace Loom.Engine.Context
{
    public class RagManager
    {
        private readonly LlmInvocation _invocation;

        public RagManager(LlmInvocation invocation)
        {
            _invocation = invocation;
        }

        /// <summary>
        /// Check if RAG context has data
        /// </summary>
        public bool HasContext => _invocation.Rag.RetrievedChunks.Any();

        /// <summary>
        /// Adds a set of retrieved results to the RAG context
        /// </summary>
        /// <param name="chunks">The retrieved text chunks</param>
        /// <param name="clearExisting">If true, clears the previous context</param>
        public void AddResults(IEnumerable<RagChunk> chunks, bool clearExisting = false)
        {
            if (clearExisting) _invocation.Rag.RetrievedChunks.Clear();

            var filtered = new List<RagChunk>();

            foreach (var chunk in chunks)
            {
                if (!string.IsNullOrWhiteSpace(chunk.Text))
                    filtered.Add(chunk);
            }

            var validChunks = filtered.OrderByDescending(c => c.Score);
            _invocation.Rag.RetrievedChunks.AddRange(validChunks);
        }

        /// <summary>
        /// Set strategy injection at runtime
        /// </summary>
        /// <param name="strategy"></param>
        public void SetStrategy(InjectionStrategy strategy)
        {
            _invocation.Rag.Strategy = strategy;
        }

    }
}
